lms/djangoapps/courseware/models.py

```diff
@@ -230,3 +230,20 @@ class Meta:
 
     def __unicode__(self):
         return "[OCGLog] %s: %s" % (self.course_id.to_deprecated_string(), self.created)  # pylint: disable=no-member
+
+
+class StudentFieldOverride(models.Model):
+    """
+    Holds the value of a specific field overriden for a student.  This is used
+    by the code in the `courseware.student_field_overrides` module to provide
+    overrides of xblock fields on a per user basis.
+    """
+    course_id = CourseKeyField(max_length=255, db_index=True)
+    location = LocationKeyField(max_length=255, db_index=True)
+    student = models.ForeignKey(User, db_index=True)
+
+    class Meta:   # pylint: disable=missing-docstring
+        unique_together = (('course_id', 'location', 'student'),)
+
+    field = models.CharField(max_length=255)
+    value = models.TextField(default='null')
```

Comment:
Okay, moving some of the points back here then:

1) Please have a composite index on ``('student', 'course_id', 'location'),)` in the migration. As modeled, if there are many rows in the database and we do a query for all of a student's overrides in a given course, it will use the individual index on `student` and possibly do an intersection with the `course_id` index -- though I'm not sure how that would play out as the table gets large, given that the `course_id` will have a much lower cardinality. With a composite index of `(student, course_id, location)`, a query for all of a particular student's course overrides would just hit that one index. It's not a big issue with the size of data involved with IDDE, but it could become important if this facility were more commonly used.

2) Could you please add timestamps (inherit from `model_utils.TimeStampedModel`)? (In #5802, @chrisrossi replied that this would be fine.)

3) Please add `field` to the unique index. This is for data integrity, not for performance. We can get into a situation where slow request processing and multiple browser clicks can cause multiple gunicorn workers to attempt to add the same row twice (this happens a lot on `StudentModule`). That would cause field override setting to permanently break for that user on that field because every subsequent attempt to call `.get()/.get_or_create()` on it would fail with a `MultipleObjectsReturned` error. If the unique constraint is extended to include `field`, then one of create attempts would throw an `IntegrityError`, and subsequent `.get()` attempts would be fine.